src/grafts.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "grafts.h"
+
+#include "oidarray.h"
+#include "parse.h"
+
+struct git_grafts {
+	/* Map of `git_commit_graft`s */
+	git_oidmap *commits;
+
+	/* File backing the graft. NULL if it's an in-memory graft */
+	char *path;
+	git_oid path_checksum;
+};
+
+int git_grafts_new(git_grafts **out)
+{
+	git_grafts *grafts;
+
+	grafts = git__calloc(1, sizeof(*grafts));
+	GIT_ERROR_CHECK_ALLOC(grafts);
+
+	if ((git_oidmap_new(&grafts->commits)) < 0) {
+		git__free(grafts);
+		return -1;
+	}
+
+	*out = grafts;
+	return 0;
+}
+
+int git_grafts_from_file(git_grafts **out, const char *path)
+{
+	git_grafts *grafts = NULL;
+	int error;
+
+	if ((error = git_grafts_new(&grafts)) < 0)
+		goto error;
+	grafts->path = git__strdup(path);
+	if ((error = git_grafts_refresh(grafts)) < 0)
+		goto error;
+
+	*out = grafts;
+error:
+	if (error < 0)
+		git_grafts_free(grafts);
+	return error;
+}
+
+void git_grafts_free(git_grafts *grafts)
+{
+	if (!grafts)
+		return;
+	git__free(grafts->path);
+	git_grafts_clear(grafts);
+	git_oidmap_free(grafts->commits);
+	git__free(grafts);
+}
+
+void git_grafts_clear(git_grafts *grafts)
+{
+	git_commit_graft *graft;
+
+	assert(grafts);
+
+	git_oidmap_foreach_value(grafts->commits, graft, {
+		git__free(graft->parents.ptr);
+		git__free(graft);
+	});
+
+	git_oidmap_clear(grafts->commits);
+}
+
+int git_grafts_refresh(git_grafts *grafts)
+{
+	git_buf contents = GIT_BUF_INIT;
+	int error, updated = 0;
+
+	if (!grafts->path)
+		return 0;
+
+	error = git_futils_readbuffer_updated(&contents, grafts->path,
+					      &grafts->path_checksum, &updated);
+	if (error < 0 || error == GIT_ENOTFOUND || !updated) {
```

Comment:
This is the kind of thing where I'm wondering "how would it work *after* I add the write-side of it" (ie. the actual part where the grafts are updated from the server's view). Swallowing `ENOTFOUND` would make an externally-triggered delete of the graft file (eg. an out-of-process unshallow takes place which just `unlink()` the file, stuff like that 😉). I haven't checked the semantics of `readbuffer_updated` in this case…

Reply:
If it's removed then it would return `GIT_ENOTFOUND`. But in fact, we should clear the grafts' contents in that case, which I missed out for now, so that definitely needs to be added